src/main/java/net/oauth/jsontoken/AbstractJsonTokenParser.java

```diff
@@ -0,0 +1,222 @@
+      issuedAt = new Instant(0);
+    }
+
+    if (issuedAt != null && expiration == null) {
+      expiration = new Instant(Long.MAX_VALUE);
+    }
+
+    if (issuedAt != null && expiration != null) {
+      if (issuedAt.isAfter(expiration)
+          || ! clock.isCurrentTimeInInterval(issuedAt, expiration)) {
+        throw new IllegalStateException(String.format("Invalid iat and/or exp. iat: %s exp: %s "
+            + "now: %s", jsonToken.getIssuedAt(), jsonToken.getExpiration(), clock.now()));
+      }
+    }
+
+    if (checkers != null) {
+      for (Checker checker : checkers) {
+        checker.check(jsonToken.getPayloadAsJsonObject());
+      }
+    }
+  }
+
+  /**
+   * Verifies that a JSON Web Token's signature is valid.
+   *
+   * @param tokenString the encoded and signed JSON Web Token to verify.
+   * @param verifiers used to verify the signature. These usually encapsulate
+   *        secret keys.
+   */
+  public boolean signatureIsValid(String tokenString, List<Verifier> verifiers) {
+    String[] pieces = splitTokenString(tokenString);
+    byte[] signature = Base64.decodeBase64(pieces[2]);
+    String baseString = JsonTokenUtil.toDotFormat(pieces[0], pieces[1]);
+
+    boolean sigVerified = false;
+    for (Verifier verifier : verifiers) {
+      AsciiStringVerifier asciiVerifier = new AsciiStringVerifier(verifier);
+      try {
+        asciiVerifier.verifySignature(baseString, signature);
+        sigVerified = true;
+        break;
+      } catch (SignatureException e) {
+        continue;
+      }
+    }
+    return sigVerified;
+  }
+
+  /**
+   * Verifies that a JSON Web Token is not expired.
+   *
+   * @param jsonToken the token to verify
+   * @param now the instant to use as point of reference for current time
+   * @return false if the token is expired, true otherwise
+   */
+  public boolean expirationIsValid(JsonToken jsonToken, Instant now) {
+    Instant expiration = jsonToken.getExpiration();
+    if ((expiration != null) && now.isAfter(expiration)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Verifies that a JSON Web Token was issued in the past.
+   *
+   * @param jsonToken the token to verify
+   * @param now the instant to use as point of reference for current time
+   * @return false if the JWT's 'iat' is later than now, true otherwise
+   */
+  public boolean issuedAtIsValid(JsonToken jsonToken, Instant now) {
+    Instant issuedAt = jsonToken.getIssuedAt();
+    if ((issuedAt != null) && now.isBefore(issuedAt)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @param tokenString The original encoded representation of a JWT
+   * @return Three components of the JWT as an array of strings
+   */
+  private String[] splitTokenString(String tokenString) {
+    String[] pieces = tokenString.split(Pattern.quote(JsonTokenUtil.DELIMITER));
+    if (pieces.length != 3) {
+      throw new IllegalStateException("Expected JWT to have 3 segments separated by '" +
```

Comment:
Can you add @throws clauses to methods that call this to specify when IllegalStateException is thrown?

Also maybe something we should consider later is throwing a different exception. It was always a mistake for this library to throw IllegalStateException here. IllegalArgumentException would make more sense, or even a custom exception. (One could imagine adding some ErrorCode enum to SignatureException to specify what the problem was, e.g. MALFORMATTED_JWT, INVALID_EXP, INVALID_IAT, INVALID_SIG, etc.)

Reply:
Agreed that throwing IllegalStateException was a mistake here, a checked exception would make more sense. Unfortunately I can see many references in google3 that now depend on this mistake and would need to be updated if we changed the exception.

Reply:
I updated the documentation to the relevant files in this PR to show when functions throw exceptions. Even if they aren't explicitly declared, they will be shown in the documentation. For other files where this is updated, another PR can be opened up later on.

Reply:
Yeah, I can look into it when implementing the async version

Reply:
A possible solution to adding more descriptive exceptions is to have custom exceptions that extend whatever exception would have been thrown. eg. IllegalStateException is thrown when a JWT is malformed, so if we have a custom exception MalformedTokenException that extends IllegalStateException. Any future code that uses the library can specifically catch these custom exceptions for a more detailed view into what went wrong. For backwards compatibility, the custom thrown exceptions would just be viewed as the exception they extended.

Reply:
One downside of that is that IllegalStateException is fairly common, and people may try to catch it for other reasons, and accidentally suppress or mishandle any exceptions thrown here.